Loop/View Models/BolusEntryViewModel.swift

```diff
@@ -0,0 +1,362 @@
+    let potentialCarbEntry: NewCarbEntry?
+    let selectedCarbAbsorptionTimeEmoji: String?
+
+    @Published var recommendedBolus: HKQuantity?
+    @Published var enteredBolus = HKQuantity(unit: .internationalUnit(), doubleValue: 0)
+    private var isInitiatingSaveOrBolus = false
+
+    @Published var activeAlert: Alert?
+
+    private let dataManager: DeviceDataManager
+    private let log = OSLog(category: "BolusEntryViewModel")
+    private var cancellables: Set<AnyCancellable> = []
+
+    let chartManager: ChartsManager = {
+        let predictedGlucoseChart = PredictedGlucoseChart()
+        predictedGlucoseChart.glucoseDisplayRange = BolusEntryViewModel.defaultGlucoseDisplayRange
+        return ChartsManager(colors: .default, settings: .default, charts: [predictedGlucoseChart], traitCollection: .current)
+    }()
+
+    static let defaultGlucoseDisplayRange = HKQuantity(unit: .milligramsPerDeciliter, doubleValue: 60)...HKQuantity(unit: .milligramsPerDeciliter, doubleValue: 200)
+
+    init(
+        dataManager: DeviceDataManager,
+        originalCarbEntry: StoredCarbEntry? = nil,
+        potentialCarbEntry: NewCarbEntry? = nil,
+        selectedCarbAbsorptionTimeEmoji: String? = nil
+    ) {
+        self.dataManager = dataManager
+        self.originalCarbEntry = originalCarbEntry
+        self.potentialCarbEntry = potentialCarbEntry
+        self.selectedCarbAbsorptionTimeEmoji = selectedCarbAbsorptionTimeEmoji
+
+        NotificationCenter.default
+            .publisher(for: .LoopDataUpdated, object: dataManager.loopManager)
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] _ in self?.update() }
+            .store(in: &cancellables)
+
+        $enteredBolus
+            .removeDuplicates()
+            .debounce(for: .milliseconds(300), scheduler: RunLoop.main)
+            .sink { [weak self] _ in
+                self?.dataManager.loopManager.getLoopState { manager, state in
+                    self?.updatePredictedGlucoseValues(from: state)
+                }
+            }
+            .store(in: &cancellables)
+
+        update()
+    }
+
+    var isBolusRecommended: Bool {
+        guard let recommendedBolus = recommendedBolus else {
+            return false
+        }
+
+        return recommendedBolus.doubleValue(for: .internationalUnit()) > 0
+    }
+
+    func saveCarbsAndDeliverBolus(onSuccess completion: @escaping () -> Void) {
+        guard dataManager.pumpManager != nil else {
+            activeAlert = .noPumpManagerConfigured
+            return
+        }
+
+        guard enteredBolus < maximumBolus else {
+            activeAlert = .maxBolusExceeded
+            return
+        }
+
+        guard let carbEntry = potentialCarbEntry else {
+            authenticateAndDeliverBolus(onSuccess: completion)
+            return
+        }
+
+        if originalCarbEntry == nil {
+            let interaction = INInteraction(intent: NewCarbEntryIntent(), response: nil)
```

Comment:
This API makes the action known to Siri Shortcuts, so that patterns (e.g. time of day) when carb entries happen can be used to surface the 'Add Carbs' shortcut.